src/environmentd/src/bin/environmentd/main.rs

```diff
@@ -742,7 +775,42 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
                     }))
                     .context("creating kubernetes orchestrator")?,
             );
-            let secrets_controller = Arc::clone(&orchestrator);
+            let secrets_controller: Arc<dyn SecretsController> =
+                match args.secrets_controller {
+                    SecretsControllerKind::Kubernetes => {
+                        let sc = Arc::clone(&orchestrator);
+                        let sc: Arc<dyn SecretsController> = sc;
+                        sc
+                    }
+                    SecretsControllerKind::AwsSecretsManager => {
+                        // One-time migrate any existing kubernetes secrets
+                        let kubernetes_secrets_controller = &orchestrator;
+                        let kubernetes_secrets: BTreeSet<GlobalId> = runtime.block_on(kubernetes_secrets_controller.list())?.into_iter().collect();
+
+                        let aws_secrets_controller = Arc::new(
+                            runtime.block_on(AwsSecretsController::new(
+                                args.environment_id.clone(),
+                                args.aws_secrets_controller_tags
+                                    .into_iter()
+                                    .map(|tag| (tag.key, tag.value))
+                                    .collect(),
+                            )),
+                        );
+                        let aws_secrets: BTreeSet<GlobalId> = runtime.block_on(aws_secrets_controller.list())?.into_iter().collect();
+
+                        let kubernetes_secrets_reader = kubernetes_secrets_controller.reader();
+                        let secrets_to_migrate = kubernetes_secrets.difference(&aws_secrets);
+                        for secret in secrets_to_migrate {
+                            let data = runtime.block_on(kubernetes_secrets_reader.read(secret.to_owned()))?;
```

Comment:
Do we need to worry about poorly timed crashes? What happens if envd does a partial write to the aws secret and crashes before fully writing it? I'm guessing that aws won't commit the write, so on next envd restart it'll not exist and will attempt to re-migrate.

What about concurrent envds both trying to migrate at the same time (no stash exists yet which could prevent that)? I thought I had a rare race condition, but I think it'd be ok because that race would have created an orphan aws secret in the catalog, which would be cleaned up next start, and never be exposed to the user.

Reply:
I believe that AWS will fail to commit the secret and the next envd restart should attempt to re-migrate.

If there are two envds concurrently trying to migrate at the same time, they will both be trying to set the same data in AWS. One will likely succeed in creating it, while the other should either get an already exists exception and then do an update of the object to the same value, or it will get some other error and crash. This should be fine either way, since the secret got set and whichever envd got there first should continue.